Move Rust templates to wasm32-wasip1

The Rust project has started the process of deprecating the `wasm32-wasi` target in favour of `wasm32-wasip1`.  (There is also a `wasm32-wasip2` but that isn't as mature I believe.)  The new target was introduced in Rust 1.78.  As of Rust 1.81, `wasm32-wasi` emits a deprecation notice; the target is slated to be removed in 1.84 (Jan 2025).

We therefore want to switch our Rust templates over to `wasm32-wasip1`, but to provide the least worst experience to users who have seen no reason to upgrade their perfectly good installation of Rust 1.73.

A possible way to do this is, at the same time we switch the template targets over, to add a MSRV of 1.78 to the templates' Cargo.toml files.  This should ensure that people on older versions get guided to bump their Rust versions.  However, I don't think upgrading to Rust 1.78 inherently installs the new target.  So a user who just upgraded Rust would still get an error on the new target:

```
error[E0463]: can't find crate for `core`
  |
  = note: the `wasm32-wasip1` target may not be installed
  = help: consider downloading the target with `rustup target add wasm32-wasip1`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `futures-core` (lib) due to 1 previous error
```

This isn't very lovely, but it does include the necessary guidance.

So I suggest that in time for the next minor release we:

* Bump Rust templates to `wasm32-wasip1`
* Bump Rust templates MSRV to 1.78
* Add docs calling out the name change and specifically including the error message

Are folks okay with this?  Better ideas?

--- a/templates/http-rust/content/Cargo.toml.tmpl
+++ b/templates/http-rust/content/Cargo.toml.tmpl
@@ -3,6 +3,7 @@ name = "{{project-name | kebab_case}}"
 authors = ["{{authors}}"]
 description = "{{project-description}}"
 version = "0.1.0"
+rust-version = "1.78"
 edition = "2021"
 
 [lib]

--- a/templates/http-rust/content/spin.toml
+++ b/templates/http-rust/content/spin.toml
@@ -11,8 +11,8 @@ route = "{{http-path}}"
 component = "{{project-name | kebab_case}}"
 
 [component.{{project-name | kebab_case}}]
-source = "target/wasm32-wasi/release/{{project-name | snake_case}}.wasm"
+source = "target/wasm32-wasip1/release/{{project-name | snake_case}}.wasm"
 allowed_outbound_hosts = []
 [component.{{project-name | kebab_case}}.build]
-command = "cargo build --target wasm32-wasi --release"
+command = "cargo build --target wasm32-wasip1 --release"
 watch = ["src/**/*.rs", "Cargo.toml"]

--- a/templates/redis-rust/content/Cargo.toml.tmpl
+++ b/templates/redis-rust/content/Cargo.toml.tmpl
@@ -3,6 +3,7 @@ name = "{{project-name | kebab_case}}"
 authors = ["{{authors}}"]
 description = "{{project-description}}"
 version = "0.1.0"
+rust-version = "1.78"
 edition = "2021"
 
 [lib]

--- a/templates/redis-rust/content/spin.toml
+++ b/templates/redis-rust/content/spin.toml
@@ -14,8 +14,8 @@ channel = "{{redis-channel}}"
 component = "{{project-name | kebab_case}}"
 
 [component.{{project-name | kebab_case}}]
-source = "target/wasm32-wasi/release/{{project-name | snake_case}}.wasm"
+source = "target/wasm32-wasip1/release/{{project-name | snake_case}}.wasm"
 allowed_outbound_hosts = []
 [component.{{project-name | kebab_case}}.build]
-command = "cargo build --target wasm32-wasi --release"
+command = "cargo build --target wasm32-wasip1 --release"
 watch = ["src/**/*.rs", "Cargo.toml"]

